File: day_3/solution2.py
# read the input file
def read_input(file_path: str) -> list[str]:
    lines = []
    with open(file_path, "r") as file:
        for line in file:
            lines.append(line.strip())
    return lines


def find_largest_joltage(input_data: list[str]) -> list[int]:
    results = []
    for line in input_data:
        largest_loc = -1
        temp_result = ""
        battery_len = 12
        while battery_len > 0:
            largest = 0
            for i in range(largest_loc + 1, len(line) - battery_len + 1):
                if largest < int(line[i]):
                    largest = int(line[i])
                    largest_loc = i
            temp_result += str(largest)
            battery_len -= 1
        results.append(int(temp_result))
    return results


# file_path = "./input_test.txt"
file_path = "./input.txt"
input_data = read_input(file_path)
# print(input_data)

result = find_largest_joltage(input_data)
# print(result)
print(sum(result))

File: day_4/solution_rust/src/solution_2.rs
use std::fs::File;
use std::io::{BufRead, BufReader};

fn read_input(path: &str) -> Result<Vec<Vec<char>>, std::io::Error> {
    let file = File::open(path)?;
    let reader = BufReader::new(file);
    let mut input: Vec<Vec<char>> = Vec::new();
    for line in reader.lines() {
        let chars: Vec<char> = line?.chars().collect();
        input.push(chars);
    }
    Ok(input)
}

fn check_valid(input: &[Vec<char>], x: usize, y: usize) -> bool {
    let directions = [
        (-1, -1),
        (-1, 0),
        (-1, 1),
        (0, -1),
        (0, 1),
        (1, -1),
        (1, 0),
        (1, 1),
    ];
    let mut count = 0;
    let len_x = input.len() as isize;
    let len_y = input[0].len() as isize;
    for direction in directions {
        let new_x = x as isize + direction.0;
        let new_y = y as isize + direction.1;
        if new_x < 0 || new_x >= len_x || new_y < 0 || new_y >= len_y {
            continue;
        }
        if input[new_x as usize][new_y as usize] != '.' {
            count += 1;
        }
        if count >= 4 {
            return false;
        }
    }
    true
}

fn solve(input: &mut [Vec<char>]) -> usize {
    let mut total = 0;
    let mut prev_total = 1;
    while total != prev_total {
        prev_total = total;
        // println!("Total: {}", total);
        for i in 0..input.len() {
            for j in 0..input[i].len() {
                if input[i][j] == '@' && check_valid(input, i, j) {
                    input[i][j] = '.';
                    total += 1;
                }
            }
        }
    }
    total
}

pub fn run() {
    let path = "../input.txt";
    // let path = "../input_test.txt";

    let mut input_data = match read_input(path) {
        Ok(data) => data,
        Err(e) => {
            eprintln!("Error reading input file: {}", e);
            return;
        }
    };
    // println!("Input Data: {:?}", input_data);

    let result = solve(&mut input_data);
    println!("Result: {}", result);
}
